src/cursor_manager.ts

```diff
@@ -302,34 +312,50 @@ export class CursorManager implements Disposable, NeovimRedrawProcessable, Neovi
             this.updateCursorStyle("visual");
         }
 
-        this.applySelectionChanged(textEditor, kind);
+        this.getDebouncedApplySelectionChanged(kind)(textEditor, kind);
     };
 
     // ! Need to debounce requests because setting cursor by consequence of neovim event will trigger this method
     // ! and cursor may go out-of-sync and produce a jitter
-    private applySelectionChanged = debounce(
-        async (editor: TextEditor, kind: TextEditorSelectionChangeKind | undefined) => {
-            // reset cursor style if needed
-            this.updateCursorStyle(this.main.modeManager.currentMode.name);
-
-            // wait for possible layout updates first
-            this.logger.debug(`${LOG_PREFIX}: Waiting for possible layout completion operation`);
-            await this.main.bufferManager.waitForLayoutSync();
-            // wait for possible change document events
-            this.logger.debug(`${LOG_PREFIX}: Waiting for possible document change completion operation`);
-            await this.main.changeManager.getDocumentChangeCompletionLock(editor.document);
-            this.logger.debug(`${LOG_PREFIX}: Waiting done`);
-
-            // ignore selection change caused by buffer edit
-            const selection = editor.selection;
-            const documentChange = this.main.changeManager.eatDocumentCursorAfterChange(editor.document);
-            if (documentChange && documentChange.isEqual(selection.active)) {
-                this.logger.debug(
-                    `${LOG_PREFIX}: Skipping onSelectionChanged event since it was selection produced by doc change`,
-                );
-                return;
-            }
+    private getDebouncedApplySelectionChanged = (
+        kind: TextEditorSelectionChangeKind | undefined,
+    ): DebouncedFunc<CursorManager["applySelectionChanged"]> => {
+        const debounceTime =
+            // Should use same debounce time if previous debounced func still pending
+            this.previousApplyDebounceTime !== undefined
```

Comment:
Maybe this should be made more readable by using if statement, switch statement, extracting into function (then you can return guard the previous time), or map?